nautobot/nautobot_data_validation_engine/models.py

```diff
@@ -0,0 +1,355 @@
+        # which is commonly seen on CharFields and results in a default of empty string which is unacceptable
+        # if the field is to be marked as required.
+        if model_field.null is False and not (model_field.null is False and model_field.blank is True):
+            raise ValidationError({"field": "This field is already required by default."})
+
+
+@extras_features(
+    "custom_fields",
+    "custom_links",
+    "export_templates",
+    "graphql",
+    "relationships",
+    "statuses",
+    "webhooks",
+)
+class UniqueValidationRule(ValidationRule):
+    """
+    A type of validation rule that applies a unique constraint to a given model field.
+
+    Optionally specify the max number of similar values for the field accross all model instances. Default of 1.
+    """
+
+    field = models.CharField(
+        max_length=50,
+    )
+    max_instances = models.PositiveIntegerField(default=1, validators=[MinValueValidator(1)])
+
+    clone_fields = ["enabled", "content_type", "max_instances", "error_message"]
+
+    class Meta:
+        """Model metadata for the UniqueValidationRule model."""
+
+        ordering = ("name",)
+        unique_together = [["content_type", "field"]]
+
+    def get_absolute_url(self, api=False):
+        """Absolute url for the instance."""
+        return reverse("nautobot_data_validation_engine:uniquevalidationrule", args=[self.id])
+
+    def clean(self):
+        """Ensure field is valid for the model and has not been blacklisted."""
+        if self.field not in [f.name for f in self.content_type.model_class()._meta.get_fields()]:
+            raise ValidationError(
+                {
+                    "field": f"Not a valid field for content type {self.content_type.app_label}.{self.content_type.model}."
+                }
+            )
+
+        blacklisted_field_types = (
+            models.Manager,
+            models.fields.related.RelatedField,
+            models.ManyToManyField,
+        )
+
+        model_field = self.content_type.model_class()._meta.get_field(self.field)
+
+        if self.field.startswith("_") or not model_field.editable or isinstance(model_field, blacklisted_field_types):
+            raise ValidationError({"field": "This field's type does not support uniqueness validation."})
+
+        if getattr(model_field, "unique", False):
+            raise ValidationError({"field": "This field is already unique by default."})
+
+
+class DataCompliance(PrimaryModel):
+    """Model to represent the results of an audit method."""
+
+    compliance_class_name = models.CharField(max_length=CHARFIELD_MAX_LENGTH, blank=False, null=False)
+    last_validation_date = models.DateTimeField(blank=False, null=False, auto_now=True)
+    content_type = models.ForeignKey(ContentType, on_delete=models.PROTECT, blank=False, null=False)
+    object_id = models.CharField(max_length=CHARFIELD_MAX_LENGTH, blank=False, null=False)
+    validated_object = GenericForeignKey("content_type", "object_id")
```

Comment:
I *think* if we added a corresponding `GenericRelation` on `BaseModel` we'd get cascade-delete behavior for this GFK so that we wouldn't need a separate cleanup job.

Reply:
Added a `GenericRelation` on `BaseModel` and confirmed the cascading delete behavior.